compiler/rustc_passes/src/check_attr.rs

```diff
@@ -2105,7 +2105,23 @@ impl CheckAttrVisitor<'_> {
     }
 
     fn check_macro_export(&self, hir_id: HirId, attr: &Attribute, target: Target) {
-        if target != Target::MacroDef {
+        // special case when `#[macro_export]` is applied to a macro 2.0
+        if target == Target::MacroDef {
+            let (macro_definition, _) =
+                self.tcx.hir().find(hir_id).unwrap().expect_item().expect_macro();
+            let is_macro_2_0 = !macro_definition.macro_rules;
+
+            if is_macro_2_0 {
+                self.tcx.emit_spanned_lint(
+                    UNUSED_ATTRIBUTES,
+                    hir_id,
+                    attr.span,
+                    errors::MacroExport::OnDeclMacro,
+                );
+            }
+        }
```

Comment:
Not sure if this should say `macro 2.0` or rather decl macro - I'm not against changing it
```suggestion
        // special case when `#[macro_export]` is applied to a decl macro
        if target == Target::MacroDef {
            let (macro_definition, _) =
                self.tcx.hir().find(hir_id).unwrap().expect_item().expect_macro();
            let is_decl_macro = !macro_definition.macro_rules;

            if is_decl_macro {
                self.tcx.emit_spanned_lint(
                    UNUSED_ATTRIBUTES,
                    hir_id,
                    attr.span,
                    errors::MacroExport::OnDeclMacro,
                );
            }
        }
```

Reply:
From a quick seach over the code, it seems to me that the term "macro 2.0" appears more often, but in identifiers, `decl_macro` appears more often than `macro_2_0`. So ideally it would be changed in that way.

Reply:
Addressed in 327361e187b